### PR TITLE
CDC: use a single `cdc$time` value for a batch of changes

### PR DESCRIPTION
**Pull request by kbr-**
If a batch update is performed with a sequence of changes with a single timestamp, they will now show up in CDC with a single timeuuid in the `cdc$time` column, distinguished by different `cdc$batch_seq_no` values.

Fixes https://github.com/scylladb/scylla/issues/5953.

### CODE REVIEW

File: cdc/log.cc [hunk 1]
```diff
@@ -938,9 +938,11 @@ cdc::cdc_service::impl::augment_mutation_call(lowres_clock::time_point timeout,
                 auto& m = mutations[idx];
                 auto& s = m.schema();
                 if (should_split(m, *s)) {
-                    for (auto&& mm : split(m, s)) {
-                        mutations.push_back(trans.transform(mm, rs.get()));
-                    }
+                    std::vector<mutation> res;
```
Comment (haaawk) [inline cdc/log.cc]: Why can't we just push back straight to the mutations?

Comment (haaawk) [reply cdc/log.cc]: Is it because insertion to mutations can invalidate m?

File: cdc/split.cc [hunk 2]
```diff
@@ -375,13 +375,12 @@ bool should_split(const mutation& base_mutation, const schema& base_schema) {
     return false;
 }
 
-std::vector<mutation> split(const mutation& base_mutation, const schema_ptr& base_schema) {
+void for_each_change(const mutation& base_mutation, const schema_ptr& base_schema, std::function<void(mutation)> f) {
```
Comment (haaawk) [inline cdc/split.cc]: Could we use noncopyable_function here to avoid allocation?

File: cdc/log.cc (hunk 1 above)
Comment (haaawk) [reply cdc/log.cc]: It seems that base_mutation is only used in for_each_change before the call to f. If you store partition key from base_mutation in a variable you should be able to insert mm straight to mutations.

File: test/boost/cdc_test.cc [hunk 3]
```diff
@@ -1126,6 +1126,13 @@ SEASTAR_THREAD_TEST_CASE(test_change_splitting) {
             }
         }
 
+        {
+            auto result = get_result({int32_type},
+                "select \"cdc$batch_seq_no\" from ks.t_scylla_cdc_log where pk = 0 and ck = 1 allow filtering");
+            std::vector<std::vector<data_value>> expected = {{int32_t(0)}, {int32_t(1)}, {int32_t(2)}, {int32_t(3)}};
+            BOOST_REQUIRE_EQUAL(expected, result);
+        }
+
```
Comment (piodul) [inline test/boost/cdc_test.cc]: Can we convert it to just `BOOST_REQUIRE_EQUAL`, just like in line 1181?

File: test/boost/cdc_test.cc [hunk 4]
```diff
@@ -1126,6 +1126,13 @@ SEASTAR_THREAD_TEST_CASE(test_change_splitting) {
             }
```
Comment (piodul) [inline test/boost/cdc_test.cc]: Can we convert it to just `BOOST_REQUIRE_EQUAL`, just like in line 1181?

Comment (haaawk) [reply test/boost/cdc_test.cc]: I checked and we can't. The order is still arbitrary.

**Comment by haaawk** [comment]
I rebased, resolved conflicts, addressed all the comments and made the PR bisectable -> https://github.com/scylladb/scylla/pull/5961

**Pull request closed without merging**